### PR TITLE
Adds support for `__slots__` assignment, refs #10801

### LINKED ISSUE #10801: __slots__ isn't effective for limiting attributes
**Bug Report**

Thank you for Mypy! It's been great help :) Recently, I want to use `__slots__` in a class definition to limit the attributes. This way I don't accidentally define attributes with similar but different names (like "bar" in one class but "baz" in another) when I copy my code to a new class. But Mypy seems to ignore it completely, even if CPython throws an `AttributeError` when executing the code.

Originally I just want to prevent definition of attributes outside `__init__`, but Mypy doesn't seem to provide an option to check this, so I thought about `__slots__` as a workaround, but unfortunately this failed too. Please point out any probable workarounds for the above issues.

**To Reproduce**

Put the following in "main.py" and first execute with python then check with mypy.

```python
class TestClass:
    __slots__ = ('bar',)
    def __init__(self) -> None:
        self.bar: int = 0
    def modify(self) -> None:
        self.baz = 42
    def __str__(self) -> str:
        return f"TestClass(bar={self.bar})"

t = TestClass()
t.modify()
print(t)
```

```
$ python main.py
$ mypy main.py
```

**Expected Behavior**

Actual Python output:

```
Traceback (most recent call last):
  File "main.py", line 11, in <module>
    t.modify()
  File "main.py", line 6, in modify
    self.baz = 42
AttributeError: 'TestClass' object has no attribute 'baz'
```

Expected Mypy output, to be consistent with the Python output above:

```
main.py:6: error: "TestClass" has no attribute "baz"
Found 1 error in 1 file (checked 1 source file)
```

**Actual Behavior**

Actual Mypy output:

```
Success: no issues found in 1 source file
```

**Your Environment** (flags, configs, and OS are probably irrelevant)

- Mypy version used: 0.910 (pypi)
- Mypy command-line flags: (None)
- Mypy configuration options from `mypy.ini` (and other config files): (None)
- Python version used: 3.7.6 (conda-forge)
- Operating system and version: macOS Catalina Version 10.15.7

### PR DESCRIPTION
**Pull request by sobolevn**
### Description

We can now detect assignment that are not matching defined `__slots__`.

Example:

```python
class A:
   __slots__ = ('a',)

class B(A):
   __slots__ = ('b',)
   def __init__(self) -> None:
       self.a = 1  # ok
       self.b = 2  # ok
       self.c = 3  # error

b: B
reveal_type(b.c)
```

Снимок экрана 2021-07-24 в 18 41 14

In runtime it produces:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 6, in __init__
AttributeError: 'B' object has no attribute 'c'
```

Closes #10801

## Test Plan

- [x] I need to test that single class with `__slots__` works
- [x] I need to test that slotted class with `__slots__` in super classes works
- [x] I need to test that a class can extend slotted class without extra `__slots__` definition
- ~~I need to test that assignment of wrong `__slots__` raise an error~~
- ~~I need to test that `__slots__` with wrong elements raise an error~~
- [x] I need to test that `__slots__` with dynamic elements are ignored
- [x] I need to test that `__slots__ = 1` outside of class definition context should be fine
- [x] I need to test `__dict__` corner case 

## Future steps

- I can also modify `attrs` plugin to include `slots` if `@attr.s(slots=True)` is passed

### CODE REVIEW

**Comment by sobolevn** [comment]
Improtant special case: `__slots__ = ('__dict__',)` which should be supported.

**Comment by sobolevn** [comment]
I am not sure why this fails: https://app.travis-ci.com/github/python/mypy/jobs/526676991

```
_____________________ testOverloadLambdaUnpackingInference _____________________

[gw0] linux -- Python 3.10.0 /home/travis/build/python/mypy/.tox/py/bin/python

data: /home/travis/build/python/mypy/test-data/unit/check-overloading.test:4002:

SystemExit: 1

----------------------------- Captured stderr call -----------------------------

The typed_ast package is not installed.

For Python 2 support, install mypy using `python3 -m pip install "mypy[python2]"`Alternatively, you can install typed_ast with `python3 -m pip install typed-ast`.
```

It does not look related.

**Comment by sobolevn** [comment]
@hauntsaninja @JelleZijlstra friendly ping. Any chance to get this PR reviewed? 🙂

File: mypy/semanal.py [hunk 1]
```diff
@@ -3353,6 +3354,62 @@ def process__deletable__(self, s: AssignmentStmt) -> None:
...
+            # We understand `__slots__` defined as string, tuple, list, set, and dict:
+            if not isinstance(s.rvalue, (StrExpr, ListExpr, TupleExpr, SetExpr, DictExpr)):
+                # For example, `__slots__` can be defined as a variable,
+                # we don't support it for now.
+                return
+
+            if any(p.slots is None for p in self.type.mro[1:-1]):
+                # At least one type in mro (excluding `self` and `object`)
+                # does not have concrete `__slots__` defined. Ignoring.
+                return
```
Comment (A5rocks) [inline mypy/semanal.py:3377]: Should this maybe leave a note or something? This sounds like something that is a bug the majority of the time, but... not sure.

(of course, I don't know the philosophy behind what mypy warns about and what it doesn't, so yeah)

File: test-data/unit/check-slots.test [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+class A:
+    __slots__ = __fields__ = ("a", "b")
+    def __init__(self) -> None:
+        self.x = 1
+        self.y = 2
+[builtins fixtures/tuple.pyi]
+
+
+
+[case testSlotsDefinitionMutlipleVars2]
```
Comment (A5rocks) [inline test-data/unit/check-slots.test:81]: *Multiple

File: test-data/unit/check-slots.test [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+[case testSlotsDefinitionOutsideOfClass]
+__slots__ = ("a", "b")
+class A:
+    def __init__(self) -> None:
+        self.x = 1
+        self.y = 2
+[builtins fixtures/tuple.pyi]
+
+
+[case testSlotsDefinitionMutlipleVars1]
```
Comment (A5rocks) [inline test-data/unit/check-slots.test:71]: *Multiple

File: test-data/unit/check-slots.test [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        self.y = 2
+[builtins fixtures/tuple.pyi]
+
+
+[case testSlotsAssignmentEmptySlots]
+class A:
+    __slots__ = ()
+    def __init__(self) -> None:
+        self.a = 1  # E: Trying to assign name "a" that is not in "__slots__" of type "__main__.A"
+        self.b = 2  # E: Trying to assign name "b" that is not in "__slots__" of type "__main__.A"
```
Comment (A5rocks) [inline test-data/unit/check-slots.test:95]: I'm not sure what the best behaviour in this case would be, I've seen this on the pyright issue tracker:

https://github.com/microsoft/pyright/issues/2039

Comment (sobolevn) [reply test-data/unit/check-slots.test:95]: Yeap, interesting corner case. On the other hand, I can imagine the opposite: some type might define empty slots to explicitly be stateless.

File: test-data/unit/check-slots.test (hunk 3 above)
Comment (A5rocks) [reply test-data/unit/check-slots.test]: I'm not sure how to mark conversations as resolved :P

File: test-data/unit/check-slots.test (hunk 4 above)
Comment (A5rocks) [reply test-data/unit/check-slots.test:95]: True yep. It's all about the tradeoffs. A mypy dev would have to choose about this, I guess.

**Review by A5rocks — APPROVED** [review]
Just a while ago I came across this behaviour of ignoring `__slots__` and it didn't seem very intuitive, thanks for doing this!

**Comment by sobolevn** [comment]
@ilevkivskyi @JukkaL Hi! Any chance to push this forward? 🙂

**Review by ethanhs — COMMENTED** [review]
I think this looks pretty reasonable overall. Just a couple of questions/comments

**Comment by ethanhs** [comment]
Based on the mypy_primer output, it seems there are also a couple of cases that should be handled:

- `__slots__` is `Any` or `list[Any]`
- the rvalue is `tuple()` or similar for dict, list, etc.

It also seems to report a few true errors :)

File: mypy/checker.py [hunk 5]
```diff
@@ -2086,6 +2086,19 @@ def accept_items(e: Expression) -> None:
     def check_assignment(self, lvalue: Lvalue, rvalue: Expression, infer_lvalue_type: bool = True,
                          new_syntax: bool = False) -> None:
         """Type check a single assignment: lvalue = rvalue."""
+        if isinstance(lvalue, MemberExpr) and lvalue.node:
+            name = lvalue.node.name
+            inst = get_proper_type(self.expr_checker.accept(lvalue.expr))
+            if (isinstance(inst, Instance) and
```
[the remark this answers is not in the document]
Comment (sobolevn) [reply mypy/checker.py:2092]: I will add a test case for this, thank you. It actually was not consistent with Python's runtime behavior:

```python
class A:
    __slots__ = ('a',)
    b = 4

    def __init__(self) -> None:
        self.a = 1
        self.b = 2  # error here in runtime

A()
# AttributeError: 'A' object attribute 'b' is read-only
```

I will change the code to make this an error with `mypy` as well.

File: test-data/unit/check-slots.test [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+[case testSlotsAssignmentWithoutSuperSlots]
+class A:
+    pass  # no slots
+class B(A):
+    __slots__ = ("a", "b")
+
+    def __init__(self) -> None:
+        self.a = 1
+        self.b = 2
+        self.missing = 3
```
[the remark this answers is not in the document]
Comment (sobolevn) [reply test-data/unit/check-slots.test:180]: This is how Python works in runtime. If any base class has `__dict__` that the child class itself would have `__dict__`. So, when using `B(A)` you will implicitly have `__dict__` in place. Try this:

```python
class A:
    pass  # no slots

class B(A):
    __slots__ = ("a", "b")

    def __init__(self) -> None:
        self.a = 1
        self.b = 2
        self.missing = 3

b = B()
b.a = 1
b.b = 2
b.missing = 3
b.extra = 4  # E: "B" has no attribute "extra"
```

But, without `A` super class it would be different:

```python
class B:
    __slots__ = ("a", "b")

    def __init__(self) -> None:
        self.a = 1
        self.b = 2
        self.missing = 3  # AttributeError: 'B' object has no attribute 'missing'

b = B()
```

[the remark this answers is not in the document]
Comment (sobolevn) [reply test-data/unit/check-slots.test:180]: I think, that I need to add this to the docs of `mypy`. Do you agree?
What would be the best place for it? Common issues?

**Comment by sobolevn** [comment]
> __slots__ is Any or list[Any]

Like just `__slots__ = returns_any()`?

> the rvalue is tuple() or similar for dict, list, etc.

Will do! 👍

**Comment by sobolevn** [comment]
Ok, this reports too many false positives. I will fix that + add more tests for `@property.setter`

**Comment by sobolevn** [comment]
One more important case I've missed: `__setattr__` support. Adding tests for it as well.

**Comment by sobolevn** [comment]
Now, let's look on cases highlighted by `mypy_primer`:

1. `dulwich`. Seems like a bug, I've opened a PR.
2. `parso`. It looks valid: https://github.com/davidhalter/parso/blob/master/parso/tree.py#L374-L385 `NodeOrLeaf` does not have `parent` field, that's true. I will report a bug for `parso`
3. `bidict`. Looks like a false-positive. It uses `nxt = property(a, b)`. I've missed this form in tests https://github.com/jab/bidict/blob/56ccfad1d9577c39828f43f8d29f39ccc0bdbb62/bidict/_orderedbase.py#L83 I will fix it today.

**Comment by ethanhs** [comment]
@sobolevn wow that's great progress in the mypy_primer output, thank you!

**Comment by sobolevn** [comment]
Looks like that `mypy_primer` now detects a single correct bug from https://github.com/python/mypy/pull/10864#issuecomment-919581484 🎉 🎉

**Comment by sobolevn** [comment]
@ethanhs any other corner-cases you can think of? 🤔

File: test-data/unit/check-slots.test (hunk 6 above)
[the remark this answers is not in the document]
Comment (ethanhs) [reply test-data/unit/check-slots.test:213]: I would probably add it to class basics: https://mypy.readthedocs.io/en/stable/class_basics.html#class-basics

**Comment by ethanhs** [comment]
> 
> 
> @ethanhs any other corner-cases you can think of? 🤔

Not really, this seems to cover things pretty well!

**Comment by sobolevn** [comment]
> Not really, this seems to cover things pretty well!

Anyways, I am here to help / fix any problems.
The last question: should we hide this behind a configuration flag?

> Do you want to add the docs change in a separate PR?

@ethanhs yes, I need to think about good places to fit it 👍

**Comment by ethanhs** [comment]
>  The last question: should we hide this behind a configuration flag?

I don't think it should be. It seems to be a useful check and with the most recent changes I don't think it should raise too many false positives, and we have so many options already...

**Comment by sobolevn** [comment]
@ethanhs agreed. Anything else I need to do to get this merged? 🙂

**Review by ethanhs — APPROVED** [review]

**Comment by ethanhs** [comment]
Thanks!

**Comment by sobolevn** [comment]
Thanks, @ethanhs! I will submit a PR with docs today 🙂
